src/shell/shell_cmds.cpp

```diff
@@ -740,6 +741,74 @@ void DOS_Shell::CMD_DIR(char * args) {
 	dos.dta(save_dta);
 }
 
+void DOS_Shell::CMD_LS(char *args)
+{
+	HELP("LS");
+
+	RealPt save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	DOS_DTA dta(dos.dta());
+
+	const std::string pattern = to_search_pattern(args);
+	bool ret = DOS_FindFirst(pattern.c_str(), 0xffff & ~DOS_ATTR_VOLUME);
+	if (!ret) {
+		WriteOut(MSG_Get("SHELL_CMD_LS_PATH_ERR"), trim(args));
+		dos.dta(save_dta);
+		return;
+	}
+
+	std::vector<DtaResult> results;
+	// reserve space for as many as we can fit into a single memory page
+	// nothing more to it; make it larger if necessary
+	results.reserve(4096 / sizeof(DtaResult));
+
+	do {
+		DtaResult result;
+		dta.GetResult(result.name, result.size, result.date,
+		              result.time, result.attr);
+		results.push_back(result);
+	} while ((ret = DOS_FindNext()) == true);
+
+	size_t w_count = 0;
+	auto maybe_print_nl = [&]() {
+		if (w_count % 5 == 0)
+			WriteOut("\n");
+	};
+
+	for (const auto &entry : results) {
+		std::string name = entry.name;
+		const bool is_dir = entry.attr & DOS_ATTR_DIRECTORY;
+
+		if (name == "." || name == "..")
+			continue;
+
+		if (is_dir)
+			upcase(name);
+		else
+			lowcase(name);
+
+		if (is_dir) {
+			WriteOut("\033[34;1m%-15s\033[0m", name.c_str());
+			++w_count;
+			maybe_print_nl();
+			continue;
+		}
+
+		const bool is_executable = ends_with(".exe", name) ||
+		                           ends_with(".bat", name) ||
+		                           ends_with(".com", name);
+
+		if (is_executable)
+			WriteOut("\033[32;1m%-15s\033[0m", name.c_str());
+		else
+			WriteOut("%-15s", name.c_str());
+
+		++w_count;
+		maybe_print_nl();
```

Comment:
Might collapse to:

``` c++
        if (is_dir) {
            upcase(name);
            WriteOut("\033[34;1m%-15s\033[0m", name.c_str());
        } else {
            lowcase(name);
            const bool is_executable = ends_with(".exe", name) ||
                                       ends_with(".bat", name) ||
                                       ends_with(".com", name);
            if (is_executable)
                WriteOut("\033[32;1m%-15s\033[0m", name.c_str());
            else
                WriteOut("%-15s", name.c_str());
        }
        ++w_count;
        maybe_print_nl();
```

Reply:
This eliminates the need for two calls to the lambda-newline function, so perhaps that could be put directly at end here too.

Reply:
I changed the code a little bit, but now I need to remember the reason why I added this lambda exactly ;)

I think I wanted to have an option for more `continue` statements in the loop, but maybe it's no longer a valid requirement.

At one point I wanted to implement `/a` or `-a` support to emulate hidden files Unix-style (for names starting with a dot), but dropped this idea (because DOSBox removes leading dot from long file names). Perhaps I had other reasons (nor marking this as resolved yet, I'll try to improve it).